devmand/gateway/src/devmand/channels/cli/QueuedCli.cpp

```diff
@@ -65,46 +90,84 @@ Future<string> QueuedCli::executeSomething(
     const Command& cmd,
     const string& prefix,
     function<Future<string>()> innerFunc) {
-  MLOG(MDEBUG) << "[" << id << "] " << prefix << "('" << cmd << "') called";
+  MLOG(MDEBUG) << "[" << id << "] " << prefix << " adding to queue ('" << cmd
+               << "')";
   shared_ptr<Promise<string>> promise = std::make_shared<Promise<string>>();
   QueueEntry queueEntry;
   queueEntry.obtainFutureFromCli = move(innerFunc);
   queueEntry.promise = promise;
   queueEntry.command = cmd.toString();
   queueEntry.loggingPrefix = prefix;
   queue.enqueue(move(queueEntry));
-  triggerDequeue();
-  return promise->getFuture(); // TODO: check lifetime
+  if (!isProcessing) {
+    triggerDequeue();
+  }
+  return promise->getFuture();
 }
 
 /*
  * Start queue reading on consumer thread if queue contains new items.
  * It is safe to call this method anytime, it is thread safe.
  */
 void QueuedCli::triggerDequeue() {
+  if (shutdown)
+    return;
   // switch to consumer thread
-  via(serialExecutorKeepAlive, [=]() {
-    MLOG(MDEBUG) << "[" << id << "] "
-                 << "isProcessing:" << this->isProcessing;
+  via(serialExecutorKeepAlive, [dis = shared_from_this()]() {
```

Comment:
It might help readability if the entire top level lambda was extracted into a function. And maybe same for the positiveResponse and negativeResponse paths below

Reply:
Adding TODO